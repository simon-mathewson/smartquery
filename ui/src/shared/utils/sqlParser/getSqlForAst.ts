import type NodeSqlParser from 'node-sql-parser';

import { getParserOptions, getSqlParser } from './utils';
import type { Connection } from '@/types/connection';

export const getSqlForAst = async (ast: NodeSqlParser.AST, engine: Connection['engine']) => {
  const sqlParser = await getSqlParser();
  return sqlParser.sqlify(ast, getParserOptions(engine));
};
