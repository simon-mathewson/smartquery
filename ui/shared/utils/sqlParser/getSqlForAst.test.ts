import type NodeSqlParser from 'node-sql-parser';
import { describe, expect, test } from 'vitest';
import { getSqlForAst } from './getSqlForAst';

const ast = {
  columns: [
    {
      as: null,
      expr: {
        column: '*',
        table: null,
        type: 'column_ref',
      },
    },
  ],
  distinct: null,
  from: [
    {
      as: null,
      db: null,
      table: 'users',
    },
    {
      as: null,
      db: null,
      join: 'INNER JOIN',
      on: {
        left: {
          column: 'id',
          table: 'users',
          type: 'column_ref',
        },
        operator: '=',
        right: {
          column: 'user_id',
          table: 'posts',
          type: 'column_ref',
        },
        type: 'binary_expr',
      },
      table: 'posts',
    },
  ],
  groupby: {
    columns: null,
    modifiers: [],
  },
  having: null,
  limit: {
    seperator: '',
    value: [
      {
        type: 'number',
        value: 1,
      },
    ],
  },
  options: null,
  orderby: [
    {
      expr: {
        column: 'id',
        table: 'users',
        type: 'column_ref',
      },
      type: 'ASC',
    },
  ],
  type: 'select',
  where: {
    left: {
      column: 'id',
      table: 'users',
      type: 'column_ref',
    },
    operator: '=',
    right: {
      type: 'number',
      value: 1,
    },
    type: 'binary_expr',
  },
  with: null,
} satisfies NodeSqlParser.AST;

describe('getSqlForAst generates SQL for AST', () => {
  test('MySQL', async () => {
    expect(await getSqlForAst(ast, { engine: 'mysql' })).toMatchSnapshot();
  });

  test('PostgreSQL', async () => {
    expect(await getSqlForAst(ast, { engine: 'postgres' })).toMatchSnapshot();
  });
});
