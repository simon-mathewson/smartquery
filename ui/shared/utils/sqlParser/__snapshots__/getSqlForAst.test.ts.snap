// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getSqlForAst generates SQL for AST > MySQL 1`] = `
"SELECT
  *
FROM
  \`users\`
  INNER JOIN \`posts\` ON \`users\`.\`id\` = \`posts\`.\`user_id\`
WHERE
  \`users\`.\`id\` = 1
ORDER BY
  \`users\`.\`id\` ASC
LIMIT
  1"
`;

exports[`getSqlForAst generates SQL for AST > PostgreSQL 1`] = `
"SELECT
  *
FROM
  "users"
  INNER JOIN "posts" ON "users"."id" = "posts"."user_id"
WHERE
  "users"."id" = 1
ORDER BY
  "users"."id" ASC
LIMIT
  1"
`;
